src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -798,17 +799,47 @@ CQueryMutators::RunExtractAggregatesMutator(Node *node,
 
 		GPOS_ASSERT(IsA(old_sublink->subselect, Query));
 
-		new_sublink->subselect = gpdb::MutateQueryOrExpressionTree(
-			old_sublink->subselect,
-			(MutatorWalkerFn) RunExtractAggregatesMutator, (void *) context,
-			0  // mutate into cte-lists
-		);
+		// One need to call the Query mutator for subselect and take into
+		// account that SubLink can be multi-level. Therefore, the
+		// context->m_current_query_level must be modified properly
+		// while diving into such nested SubLink.
+		new_sublink->subselect =
+			RunExtractAggregatesMutator(old_sublink->subselect, context);
 
 		context->m_current_query_level--;
 
 		return (Node *) new_sublink;
 	}
 
+	if (IsA(node, Query))
+	{
+		Query *query = gpdb::MutateQueryTree(
+			(Query *) node,
+			(MutatorWalkerFn) CQueryMutators::RunExtractAggregatesMutator,
+			context,
+			QTW_IGNORE_RT_SUBQUERIES  // flag -- do not mutate range table entries
+		);
+
+		List *rtable = query->rtable;
+		ListCell *lc = NULL;
+		ForEach(lc, rtable)
```

Comment:
I suggest to replace lines 823-825 with
```
ListCell *lc;
ForEach(lc, query->rtable)
```

lc value from 824 line is never used, because the value is rewritten in ForEach. The rtable variable is used only once, so we can use query->rtable instead of it.

Reply:
Done

Reply:
Wait, no, its wrong to change the pointer, the query is later returned.

Reply:
Probably It is better to add a NULL check for rte
```
RangeTblEntry *rte = (RangeTblEntry *) lfirst(lc);
GPOS_ASSERT(NULL != rte);
```

Reply:
> Probably It is better to add a NULL check for rte
No, it's ok. If NULL will be inside for, the loop won't start and the query will be returned. Everything is fine.
```
#define ForEach(cell, l) \
	for ((cell) = gpdb::ListHead(l); (cell) != NULL; (cell) = lnext(cell))
```